### PR TITLE
Jest Testing, Coverage, CI

### LINKED ISSUE #5: Test auth flow
After auth middleware works on the phones I have, lets add some tests

### PR DESCRIPTION
**Pull request by audiolion**
So far, got tests setup with babel, jest, ts, some sip messages fixtures, and a basic test.

todo:
- [x] coverage
- [x] finish testing auth flow

should we be adding CI to this PR or do it afterwards?

Closes #5

### CODE REVIEW

**Comment by zackify** [comment]
Nice. I think we can put the sip messages right at the top of the tests as a template string so there’s less places to look and the test has the full picture of what’s happening. Also, sure we could add circleci to this pr. Looking great so far!

**Comment by zackify** [comment]
I put the sip message right into the test. we can make a file for each test if they get too big, but i think it's better to see it all in one place. also added a resolve alias so you can do `src/parser` no matter how nested you are. we can add more aliases if we need them, but only the shared folder should imo

File: src/parser.ts [hunk 1]
```diff
@@ -24,13 +24,13 @@ const parsers: any = {
 export const parse = (message: string): Message => {
   if (process.env.DEBUG) console.log(`\n\nReceived: \n\nSIP/2.0 ${message}\n`);
 
-  let [type] = message.split(' ');
+  let [type] = message.trim().split(' ');
```
Comment (zackify) [inline src/parser.ts]: added this for the tests, but figure some phones might have spaces and break it if we dont trim

**Comment by audiolion** [comment]
I will let you review/refactor, I am not super happy with the names but I couldn't come up with anything better

File: src/parser/to-js/header.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/*
+  Breaks header contents into a js object of the form
+  { raw: <headerValue>, key: value, key: value ... }
+  where key: value are the parameters included in the
+  header, delimited by ";".
+*/
+
+export default (line: string) => {
+  const [raw, ...rest] = line.split(';');
+  if (rest.length > 0) {
```
Comment (zackify) [inline src/parser/to-js/header.ts]: can just be `if (rest.length) {`

File: src/parser/to-js/header.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    const items = rest[0].split(',');
+    let params = items.map(item => {
+      const vals = item.split('=');
+      return {[vals[0]]: vals[1]};
+    })
+    params = Object.assign({}, ...params);
+    return { raw, ...params };
```
Comment (zackify) [inline src/parser/to-js/header.ts]: No need for object.assign here, `return { ...params, raw };` should work fine (also switched so raw cant be overwritten)

File: src/parser/to-sip.ts [hunk 4]
```diff
@@ -1,10 +1,16 @@
 //Turns JS version of SIP message back to sip
 import decamelize from 'decamelize';
 
+const stringify = obj => {
+  const {raw, ...rest} = obj;
+  const params = Object.entries(rest).map(entry => `${entry[0]}=${entry[1]}`).join(',');
+  return params ? [raw, params].join(';') : raw;
+};
+
 export default ({ message, status }: { message: Message; status: string }) => {
```
Comment (zackify) [inline src/parser/to-sip.ts]: let's make a new folder, to-sip/index.ts and to-sip/headers.ts for the stringify? maybe a better name than that?

File: src/shared/auth/send-password-request.ts [hunk 5]
```diff
@@ -18,8 +18,8 @@ export default ({ socket, message }: Data, method: Method) => {
   message.algorithm = message.algorithm || 'md5';
   const headerName = method === 407 ? 'proxyAuthenticate' : 'wwwAuthenticate';
   const headerContents = `Digest ${headerize(message)}`;
-  const updatedTo = `${message.from.split(';')[0]};tag=${createTag()}`
-
+  const updatedTo = Object.assign({}, message.from);
```
Comment (zackify) [inline src/shared/auth/send-password-request.ts]: Why object.assign? you can just do {...message.from} it is the same

**Comment by zackify** [comment]
Looking really good, I think it's fine minus the minor changes I commented on! One question: what do you think about passing in the nonce functions instead of importing them wherever you have to mock the function. that would get rid of the mocks.

Right now we are doing: 
```
import createNonce from './create-noce'

export default () => {
  createNonce()
}
```

Instead, we could pass it to the functions needing it:

```
export default (createNonce) => {
  createNonce()
}
```

File: tests/index.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  expect(server).toHaveProperty('use');
+});
+
+test('server gets created and returns use', () => {
+  let server = createServer({ port: 999 });
+  server.use('test', () => true);
+  server.use(() => true);
+
+  expect(server).toHaveProperty('use');
+});
```
Comment (zackify) [inline tests/index.js]: So i got the mocks working here and the coverage is 100, but I'm not sure how to expect the jest.fn's to have been called in the mocks. if I made `let mock = jest.fn` and set the value of jest.mock to the mock variable, it doesnt work. the jest.mocks are called before the variables are set. wonder if you know how to add assertions for these.

**Comment by zackify** [comment]
All looks good to me, if you know how to add assertions on the index.js i would love to add those.

File: tests/parser/to-sip/headers.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import headers from 'src/parser/to-sip/headers';
+
+test('Headers are returned if they are strings', () => {
+  expect(headers('blah')).toEqual('blah');
+});
+
+test('Headers are converted from object to string', () => {
```
Comment (audiolion) [inline tests/parser/to-sip/headers.js]: I would add one two more tests, `headers({ raw: '1 REGISTER' })` and `headers({ raw: '1 REGISTER', tag: 123, branch: 456 })`

File: tests/index.js (hunk 6 above)
Comment (audiolion) [reply tests/index.js]: off the top of my head I am not sure, maybe we can make a note and circle back to this

File: tests/parser/to-sip/headers.js (hunk 7 above)
Comment (zackify) [reply tests/parser/to-sip/headers.js]: Good catch, did that :)

File: tests/index.js (hunk 6 above)
Comment (audiolion) [reply tests/index.js]: nice! good to know how to do this

**Comment by zackify** [comment]
Decided its good enough to merge, if we find issues we can make branches for them

**Pull request merged**